zapcore/write_syncer.go

```diff
@@ -75,6 +78,99 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	sync.Mutex
+	bufferWriter *bufio.Writer
+	ticker       *time.Ticker
+}
+
+const (
+	// defaultBufferSize sizes the buffer associated with each WriterSync.
+	defaultBufferSize = 256 * 1024
+
+	// defaultFlushInterval means the default flush interval
+	defaultFlushInterval = 30 * time.Second
+)
+
+// CloseFunc should be called when the caller exits to clean up buffers.
+type CloseFunc func() error
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (WriteSyncer, CloseFunc) {
+	if _, ok := ws.(*bufferWriterSyncer); ok {
+		// no need to layer on another buffer
+		return ws, func() error { return nil }
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if bufferSize == 0 {
+		bufferSize = defaultBufferSize
+	}
+
+	if flushInterval == 0 {
+		flushInterval = defaultFlushInterval
+	}
+
+	ticker := time.NewTicker(flushInterval)
+
+	ws = &bufferWriterSyncer{
+		bufferWriter: bufio.NewWriterSize(ws, bufferSize),
+		ticker:       ticker,
+	}
+
+	// flush buffer every interval
+	// we do not need to exit this goroutine until closefunc called explicitly
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				// the background goroutine just keep syncing
+				// until the close func is called.
+				_ = ws.Sync()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	closefunc := func() error {
+		cancel()
+		return ws.Sync()
```

Comment:
We need to stop the ticker here first as well. Otherwise we'll leak the ticker goroutine.

```suggestion
		ticker.Stop()
		cancel()
		return ws.Sync()
```

I think it's also preferable for closefunc to block until the goroutine exits
so that we can be sure of a clean exit. We can do the final flush after the
goroutine exits.

```
stopped := make(chan struct{})
go func() {
  defer close(stopped)
  for { ... }
}

closefunc := func() error {
  ticker.Stop()
  cancel()
  <-stopped
  return ws.Sync()
}
```

Reply:
Sure, we could do this through a channel.